mlcvs/cvs/unsupervised/vae.py

```diff
@@ -182,13 +182,17 @@ def training_step(self, train_batch, batch_idx):
         if 'weights' in train_batch:
             options['weights'] = train_batch['weights']
 
-        # TODO: Should we do preprocessing here?
-
         # Encode/decode.
         mean, log_variance, x_hat = self.encode_decode(x)
 
+        # Reference output (compare with a 'target' key if any, otherwise with input 'data')
+        if 'target' in train_batch:
+            x_ref = train_batch['target']
+        else:
+            x_ref = x 
+
         # Loss function.
-        loss = self.loss_fn(x, x_hat, mean, log_variance, **options)
+        loss = self.loss_fn(x_hat, x_ref, mean, log_variance, **options)
```

Comment:
I'd keep the order ``(x_ref, x_hat, ...)`` here.

The order doesn't matter when the decoder's distribution is a Gaussian like the default (which gives rise to the MSE loss), but when the decoder's distribution is not isotropic, the datapoint (``x_ref``) and the distribution's parameters (``x_hat``) are not usually interchangeable.